Cilium CLI IPsec fixes

viktor-kurchenko opened the pull request:
Cilium CLI IPsec fixes:
1. `expectedIPsecKeyCount` function fixed according to [the rules from CI](https://github.com/cilium/cilium/blob/main/.github/actions/ipsec-key-rotate/action.yaml#L41C9-L52C27).
2. Make `key-per-node` param visible to users

jschwinger233 reviewed (commented):
I think the "+4 for tunnel routing" is only expected for 1.17+. Maybe we should do some versioning control?
```go
if ciliumVersion > 1.17 {
   // new expected keys 
} else {
  // old algorithm
}
```

viktor-kurchenko committed:
cli,ipsec: remove key-per-node param

The commit removes `key-per-node` parameter as newer Cilium versions
support only IPsec key per node.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
cli,ipsec: expected key count fixed

The commit fixes ipsecExpectedKeyCount function and restricts the
functionality to the Cilium version 1.18 and higher.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

pchaigno commented on cilium-cli/encrypt/ipsec_rotate_key.go:
```diff
@@ -98,9 +90,7 @@ func keyFromSlice(parts []string) (ipsecKey, error) {
 		return ipsecKey{}, fmt.Errorf("IPsec key invalid [expected parts: 5, actual parts: %d]", len(parts))
 	}
-	spiSuffix := false
 	if strings.HasSuffix(parts[1], "+") {
```
Can't this condition go away as you did in `cipherKeyFromSlice`?

pchaigno replied:
Ah, I just saw this is done, but in the wrong commit :slightly_smiling_face:

viktor-kurchenko replied:
My bad, fixed! Thanks!

pchaigno commented on cilium-cli/encrypt/status.go:
```diff
@@ -52,6 +64,49 @@ func (s *Encrypt) PrintEncryptStatus(ctx context.Context) error {
+	resCh := make(chan res)
+	defer close(resCh)
+
+	// concurrently fetch state from each cilium pod
```
I'm wondering if this isn't a bit over-engineered. I don't think we go to such lengths anywhere else in the CLI. What's the worse that can happen if we skip this?

viktor-kurchenko replied:
In the worst case, we can have a cluster with an upgrade in progress...
I agree that it's a pretty rare scenario.
Is it better to simplify the code and read the Cilium version from the first pod?

pchaigno replied:
So they'd perform a key rotation during an upgrade. Clearly uncovered in CI, but what would break? I guess if the upgrade is from v1.17 to v1.18, then they still can't be using the old key system (we removed support in v1.17). Is there something else in v1.17 we were worried about?

viktor-kurchenko replied:
I agree but then the question is: which version should we consider if there are pods with different versions?
I mean if we simply select the first pod it might still run 1.17 (which is not supported) but the second pod might be already updated to 1.18.

We can read all the versions and select the highest one, WDYT?

pchaigno replied:
I'm suggesting to check only one pod and if it's lower than v1.18, we bail out. Is there a scenario where that can cause issues?

viktor-kurchenko replied:
Not a big deal, user can always re-run the command.
Refactored.
I've also dropped the last commit as not needed with the current logic.

pchaigno commented on cilium-cli/encrypt/encrypt.go:
```diff
@@ -54,5 +55,8 @@ func (s *Encrypt) fetchCiliumPods(ctx context.Context) ([]corev1.Pod, error) {
 	}
+	if len(pods.Items) == 0 {
+		return nil, errors.New("unable to find Cilium pods")
+	}
```
I'm guessing many parts of the CLI use this function. Are we sure we're not going to break stuff with this change in the function's behavior? Why is this change needed?

viktor-kurchenko replied:
This method is used only in 1 place: https://github.com/cilium/cilium/pull/37018/files#diff-73920c9c21ba8049dfa165ca3e9f5cf0e05ca9b84c80932b2a95937bcff0d411R31
I think it doesn't make sense to continue if there are no Cilium pods.

pchaigno replied:
Oh, I see. `Encrypt` has its own `fetchCiliumPods`. Sorry for the noise!

pchaigno reviewed (commented):
Couple questions, but it looks good otherwise :rocket:

pchaigno reviewed (approved):
Looks good to me. Thanks Viktor! :pray:
